lib/utils/packaging/unarchive_unix.go

```
@@ -0,0 +1,169 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package packaging
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+
+	"github.com/google/renameio/v2"
+	"github.com/gravitational/trace"
+)
+
+// ReplaceToolsBinaries extracts the package into `extractDir` and replaces the specified
+// applications with symlinks in the tool's directory.
+func ReplaceToolsBinaries(toolsDir string, archivePath string, extractDir string, apps []string) error {
+	switch runtime.GOOS {
+	case "darwin":
+		return replacePkg(toolsDir, archivePath, extractDir, apps)
+	default:
+		return replaceTarGz(toolsDir, archivePath, extractDir, apps)
+	}
+}
+
+// replaceTarGz un-archives the Teleport package in .tar.gz format, iterates through
+// the compressed content, and ignores everything not matching the app binaries specified
+// in the apps argument. The data is extracted to extractDir, and symlinks are created
+// in toolsDir pointing to the extractDir path with binaries.
+func replaceTarGz(toolsDir string, archivePath string, extractDir string, apps []string) error {
+	f, err := os.Open(archivePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer f.Close()
+
+	gzipReader, err := gzip.NewReader(f)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		// Skip over any files in the archive that are not in apps.
+		if !slices.ContainsFunc(apps, func(s string) bool {
+			return strings.HasSuffix(header.Name, s)
+		}) {
+			continue
+		}
+		// Verify that we have enough space for uncompressed file.
+		if err := checkFreeSpace(extractDir, uint64(header.Size)); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err = func(header *tar.Header) error {
+			dest := filepath.Join(toolsDir, strings.TrimPrefix(header.Name, "teleport/"))
```

What happens if the tarball has subpaths? E.g., `teleport/blah/tsh`?

---

I think we should also make sure that any subpaths are created (`Mkdir`) or forbid them in the `executables` list.

---

Agree, but the function should return a clear error if I pass `bin/tsh` instead of `tsh` in via `executables` list.

---

I changed logic so `execName` must only match file name in archive